src/decoder/mod.rs

```diff
@@ -620,11 +620,11 @@ impl<R: Read> Reader<R> {
         match (color_type, trns) {
             (ColorType::Indexed, _) if expand => {
                 output_buffer[..row.len()].copy_from_slice(row);
-                expand_paletted(output_buffer, info, trns)?;
+                expand_paletted(row, output_buffer, info, trns)?;
             }
             (ColorType::Grayscale | ColorType::GrayscaleAlpha, _) if bit_depth < 8 && expand => {
                 output_buffer[..row.len()].copy_from_slice(row);
-                expand_gray_u8(output_buffer, info, trns)
+                expand_gray_u8(row, output_buffer, info, trns)
```

Comment:
I was able to delete these `copy_from_slice`s and all tests passed, but I'm not sure if that's always correct. I didn't see a performance difference in benchmarks so I left them alone.

It seemed like it could be a bug if `buf.len() < channels` since `unpack_bits` wouldn't expand any data. `buf` would be left with whatever was in the scratch buffer instead of the previous row.

I'm not sure if that's a possible condition to be in by the time the code gets here or if it would have errored out earlier.

Reply:
The intention when I added those `copy_from_slice` calls is that they'd be removed once `expand_{paletted, gray_8}` were refactored (the other cases below already don't use them). It is worth double checking they're not hiding any bugs / size mismatches, but I don't think we should keep them just in case

Reply:
Pushed a commit to remove them.

Reply:
`next_raw_interlaced_row` makes sure that we have a full row and returns an error if not, so I think that makes it okay to remove these.